### DISCUSSION TITLE
Optimize Large File Upload

### DISCUSSION THREAD

**Discussion by PatrickMessierGoC** [livewire/livewire#2188] [3. Feature ideas]
# Context
We are developing a front end application for a video streaming solution. We use a laravel/livewire application to upload video files to a video backend server. Video files can be up to 2 GiB in size. The application is bundled in a container and run on a cloud provider. Cloud container nodes have limited disk IO throughput (typically 25Mb/sec for writing) and we can't change it.

## Application Stack
- nginx 1.18
- php-fpm 7.4
- laravel v8.0
- livewire v2.3.0

# Problem
When dealing with large files (>100 MB), we start seeing delays between the time the upload progress bar indicates the end of the file transfer, and the time livewire confirms the end of the processing. The delay grows linearly with the file size. For a 500 MB video file, the delay is typically around 30 secs. Note that this excludes any file processing. We are just dealing with the uploading of the file onto the web server.

# Analysis
Here is my rough understanding of what's happening when a file is uploaded in a laravel/livewire application and where it hurts for large files (identified below with **):
1. Client (browser) initiates a file upload.
1. Nginx receives a post request for the application. The connection is handed to the application through the fastcgi interface (php-fpm).
1. PHP-FPM buffers the uploaded file in memory. 
1. Once the file is fully uploaded, PHP-FPM process saves it to the web server file system at a location set by php-fpm settings (upload_tmp_dir or sys_temp_dir). 
1. ** The application copies the file from php-fpm temporary location to its own temporary location set by config/livewire.php (temporary_file_upload/directory). 
1. ** The application deletes the original temporary file. 
1. The application generates a response and returns it to Nginx.
1. Nginx sends the application response to client.

# Issue
Based on the above observations, the file is "copied" from php-fpm location to livewire temp directory instead of being "moved". This litterally doubles the time required to process the file as the file is written twice to the disk. More over, it also doubles the disk space required as the file exists in two places for a short period of time. When the application is deployed on small nodes, this become an issue (time and spacewise).

# Steps to reproduce
Monitor the content of php-fpm and livewire upload temp directories while uploading a large file (>100MB - If you don't have access to a slow VM, increase the file size). 
You will see that the file is first created in php-fpm temp directory, then a copy is created in livewire directory, then the original file is deleted from php-tmp dir. The copy-delete process can take several seconds to complete. In comparaison, if the same file is moved from one location to the other (assuming the two locations are on the same file system), then the move is typically instantaneous. 

# Feature Request
- Implement a parameter in Livewire config file to control whether uploaded files should to be either "copy-deleted" (when different file systems are used) or "moved" (when only one file system is used) from the php upload temp to the livewire upload directory and document the use cases where one option is preferable over the other.

**Comment by komen205** [comment]
Same for me. It takes some time to process 500MB videos.

**Reply by PatrickMessierGoC** [reply]
Glad to here I'm not alone in this situation. Any workarounds or best practices on your side?

**Reply by komen205** [reply]
Nope. I'm still trying to figure out something.
I guess I will have to use laravel file upload or other solution until we get to something.

**Reply by komen205** [reply]
From what I've seen. Livewire creates a temporary path where the file is stored. I think a better solution would be, after selecting the file. It'd start the upload process. So it wouldn't spend less time. @PatrickMessierGoC 
I'm not sure if we can acomplish this. Maybe we should create a issue, atleast it'd get more attention. I'm not sure....

**Reply by PatrickMessierGoC** [reply]
@komen205 If you think can reformulate this feature request into a bug, you might as well try it. Your call.

**Reply by komen205** [reply]
@PatrickMessierGoC I found a solution to increase upload time but doesn't use livewire. I used pionl/laravel-chunk-upload . I think this can be adapted to livewire.

**Comment by vigstudio** [comment]
I am also having the same problem. Users often have to upload files with a capacity of more than 1GB, waiting for a long time but reporting an error that cannot be uploaded

**Reply by vigstudio** [reply]
I think we should use chunk upload

**Comment by fflnvb** [comment]
Same Experience over here. Large Files are scratching the overall great experience that Livewire adds to Laravel. It feels slow and the Progress indicator is at 100% way too early, leading to irritation rather than orientation.

Has anyone come to proper workarounds right now? Could Medialibrary Pro be a proper alternative? Would greatly appreciate your input and experience over the past years.

**Comment by thalisvilela** [comment]
Any development on this issue guys? I just stumbled uppon the same problem... After the file is completelly uploaded, it still takes an anoying amount of time to complete the request.

**Comment by RealMrHex** [comment]
Same issue, i used FilePond as a workaround.

**Reply by younus93** [reply]
I use filepond + livewire. I always find my large files stuck at 100% for a couple of minutes before it eventually succeeds. I use both s3 for temporary and permanent file storage.

**Reply by Pushkraj19** [reply]
@younus93 same issue with me. Even 5-10 mb file stuck at 100% for 10-20 seconds. I have server with gbps internet but still same.
